accepted/wasmtime-rpc.md

```diff
@@ -0,0 +1,162 @@
+For example, `wasi:keyvalue` host interface implementations are most useful when they are able to interact with real key-value stores.
+Apart from additional maintenance burden, adding support for these concrete implementations pollutes the dependency graph of `wasmtime` itself, which in turn:
+    - increases the binary size
+    - makes `wasmtime` more susceptible to supply-chain attacks
+    - negatively affects build speeds
+- Integrations with some services may not be possible to be implemented in `wasmtime` or even as part of a Bytecode Alliance project due to licensing incompatibilities.
+
+From perspective of `wasmtime` embedders, host interface implementations are tightly coupled with `wasmtime` version and so are a significant maintenance burden, especially if maintained outside of `wasmtime` tree.
+`wasmtime` linker API as well as `wasmtime` WASI implementation are effectively the API surface that developers must target to extend `wasmtime` embeddings.
+Maintainers of, for example, Rust crates that provide host interface implementations require a release cadence synchronised with `wasmtime` with each release being a breaking change due to `wasmtime` crate version increasing the major version.
+
+Even though WebAssembly components become more and more capable over time, the need for direct access to host functionality (e.g. access to hardware devices) is unlikely to become redundant any time soon.
+
+Plugin use case requirements vary greatly - for example, while for some users running plugins as part of their WebAssembly runtime process may be acceptable, for others it may not be.
+
+See https://github.com/bytecodealliance/wasmtime/issues/7348 for additional context on the CLI use case
+
+# Proposal
+[proposal]: #proposal
+
+The proposal is to allow extending `wasmtime` CLI with out-of-tree, RPC-based host plugins and use [wRPC] as the (first, potentially out of many) RPC protocol.
+
+This would be an *addition* to existing functionality, rather than a replacement of it, which would compose with what is already available.
+
+The CLI `run` and `serve` commands would support an additional option, it seems that `-P` for `--plugin` would fit well:
+
+```
+  -P, --plugin <KEY[=VAL[,..]]>
+          Plugin-related configuration options, `-P help` to see all
+```
+
+```
+Available plugin options:
+
+  -P protocol=<protocol> --
+
+        Selects the plugin protocol to use, e.g. `wrpc+tcp` or `wrpc+unix`
+  -P target=<target> --
+
+        Specifies the protocol-specific plugin target, which provides
+        imports.
+        For `wrpc+tcp`, this is a resolvable address, e.g. `localhost:7761`
+        For `wrpc+unix`, this is a path on the file system e.g. `/tmp/wasmtime-rpc`
+  -P import=<spec>[,spec[,..]] --
+
+        Selects the interfaces to import using the plugin.
+        The value is a sequence comma-separated interface specifiers.
+        For example: `wasi:keyvalue` or `wasi:keyvalue,wasi:nn`
```

Comment:
It seems like this wouldn't allow using multiple plugins at the same time. Only a single plugin exposing multiple interfaces. Maybe a syntax like `-P import=wasi:keyvalue=wrpc+tcp://[::1]:7761 -P import=wasi:nn=wrpc+unix://run/user/1000/wasi_nn.sock` would work?

Reply:
Definitely open to suggestions here, I struggled to come up with a design, which would not feel overloaded and as I mention in the *open questions* section. I've decided to start with a singleton plugin because of the CLI design concerns and the fact that shared state (e.g. resources) gets increasingly more tricky as number of plugins grows.

Perhaps we would just need to come up with `wasmtime.toml`-or-similar config file eventually, which would define the mapping, since mapping every single interface to an endpoint on the CLI feels quite tedious if the number of interfaces is large.
That's probably not much worse than the likes of `docker run -v$path:$path -w$path`, so perhaps endpoint per-interface is fine at least to start with